From b1f0e2b8d077c3905d9b087b9c58e5a5d3c5210d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tom=C3=A1s=20Gonz=C3=A1lez?=
 <tomasagustin.gonzalezorlando@arm.com>
Date: Tue, 14 Mar 2023 11:22:54 +0000
Subject: [PATCH] net: zperf: allow TCP receiver to handle multiple
 connections

Currently the zperf_tcp_receiver can handle only one TCP connection
each time, modify the code to poll and handle multiple connections.

Take the occasion to unify the bind and listen part of the code
between ipv4 and ipv6 part using a structure introduced to handle
the multiple connections.

Issue-Id: SCM-6210
Signed-off-by: Luca Fancellu <luca.fancellu@arm.com>
Change-Id: I194add76a50b0abd3e48b1b72fb7912622646219
Upstream-Status: Backport [main:6748d588abd161f80cf307986a05df072fd23788]
---
 subsys/net/lib/zperf/zperf_tcp_receiver.c | 124 +++++++++++++---------
 1 file changed, 76 insertions(+), 48 deletions(-)

diff --git a/subsys/net/lib/zperf/zperf_tcp_receiver.c b/subsys/net/lib/zperf/zperf_tcp_receiver.c
index 25f2e0a7b68..d01a4407cac 100644
--- a/subsys/net/lib/zperf/zperf_tcp_receiver.c
+++ b/subsys/net/lib/zperf/zperf_tcp_receiver.c
@@ -34,7 +34,7 @@ LOG_MODULE_DECLARE(net_zperf, CONFIG_NET_ZPERF_LOG_LEVEL);
 #define SOCK_ID_IPV4_DATA 1
 #define SOCK_ID_IPV6_LISTEN 2
 #define SOCK_ID_IPV6_DATA 3
-#define SOCK_ID_MAX 4
+#define SOCK_ID_MAX		(CONFIG_NET_ZPERF_MAX_SESSIONS + 2)
 
 #define TCP_RECEIVER_BUF_SIZE 1500
 #define POLL_TIMEOUT_MS 100
@@ -49,6 +49,40 @@ static bool tcp_server_stop;
 static uint16_t tcp_server_port;
 static K_SEM_DEFINE(tcp_server_run, 0, 1);
 
+static int tcp_bind_listen_connection(struct zsock_pollfd *pollfd,
+									  struct sockaddr *address)
+{
+	uint16_t port;
+	int ret;
+
+	if (address->sa_family == AF_INET) {
+		port = ntohs(net_sin(address)->sin_port);
+	} else {
+		port = ntohs(net_sin6(address)->sin6_port);
+	}
+
+	ret = zsock_bind(pollfd->fd, address, sizeof(*address));
+	if (ret < 0) {
+		NET_WARN("Cannot bind IPv%d TCP port %d (%d)\n",
+			      (address->sa_family == AF_INET ? 4 : 6),
+			      port, errno);
+		goto out;
+	}
+
+	ret = zsock_listen(pollfd->fd, 1);
+	if (ret < 0) {
+		NET_WARN("Cannot listen IPv%d TCP (%d)",
+			      (address->sa_family == AF_INET ? 4 : 6),
+			      errno);
+		goto out;
+	}
+
+	pollfd->events = ZSOCK_POLLIN;
+
+out:
+	return ret;
+}
+
 static void tcp_received(const struct sockaddr *addr, size_t datalen)
 {
 	struct session *session;
@@ -101,7 +135,8 @@ static void tcp_received(const struct sockaddr *addr, size_t datalen)
 static void tcp_server_session(void)
 {
 	static uint8_t buf[TCP_RECEIVER_BUF_SIZE];
-	struct zsock_pollfd fds[SOCK_ID_MAX] = { 0 };
+	static struct zsock_pollfd fds[SOCK_ID_MAX];
+	static struct sockaddr sock_addr[SOCK_ID_MAX];
 	int ret;
 
 	for (int i = 0; i < ARRAY_SIZE(fds); i++) {
@@ -144,22 +179,16 @@ static void tcp_server_session(void)
 		NET_INFO("Binding to %s",
 			 net_sprint_ipv4_addr(&in4_addr->sin_addr));
 
-		ret = zsock_bind(fds[SOCK_ID_IPV4_LISTEN].fd,
-				 (struct sockaddr *)in4_addr_my,
-				 sizeof(struct sockaddr_in));
-		if (ret < 0) {
-			NET_ERR("Cannot bind IPv4 TCP port %d (%d)",
-				ntohs(in4_addr->sin_port), errno);
-			goto error;
-		}
+		memcpy(net_sin(&sock_addr[SOCK_ID_IPV4_LISTEN]), in4_addr,
+				sizeof(struct sockaddr_in));
 
-		ret = zsock_listen(fds[SOCK_ID_IPV4_LISTEN].fd, 1);
+		ret = tcp_bind_listen_connection(
+				&fds[SOCK_ID_IPV4_LISTEN],
+				&sock_addr[SOCK_ID_IPV4_LISTEN]);
 		if (ret < 0) {
-			NET_ERR("Cannot listen IPv4 TCP (%d)", errno);
 			goto error;
 		}
 
-		fds[SOCK_ID_IPV4_LISTEN].events = ZSOCK_POLLIN;
 	}
 
 	if (IS_ENABLED(CONFIG_NET_IPV6)) {
@@ -199,18 +228,13 @@ static void tcp_server_session(void)
 		NET_INFO("Binding to %s",
 			 net_sprint_ipv6_addr(&in6_addr->sin6_addr));
 
-		ret = zsock_bind(fds[SOCK_ID_IPV6_LISTEN].fd,
-				 (struct sockaddr *)in6_addr,
-				 sizeof(struct sockaddr_in6));
-		if (ret < 0) {
-			NET_ERR("Cannot bind IPv6 TCP port %d (%d)",
-				ntohs(in6_addr->sin6_port), errno);
-			goto error;
-		}
+		memcpy(net_sin6(&sock_addr[SOCK_ID_IPV6_LISTEN]), in6_addr,
+				sizeof(struct sockaddr_in6));
 
-		ret = zsock_listen(fds[SOCK_ID_IPV6_LISTEN].fd, 1);
+		ret = tcp_bind_listen_connection(
+				&fds[SOCK_ID_IPV6_LISTEN],
+				&sock_addr[SOCK_ID_IPV6_LISTEN]);
 		if (ret < 0) {
-			NET_ERR("Cannot listen IPv6 TCP (%d)", errno);
 			goto error;
 		}
 
@@ -220,7 +244,6 @@ static void tcp_server_session(void)
 	NET_INFO("Listening on port %d", tcp_server_port);
 
 	while (true) {
-		struct sockaddr addr_ipv4, addr_ipv6;
 
 		ret = zsock_poll(fds, ARRAY_SIZE(fds), POLL_TIMEOUT_MS);
 		if (ret < 0) {
@@ -237,8 +260,6 @@ static void tcp_server_session(void)
 		}
 
 		for (int i = 0; i < ARRAY_SIZE(fds); i++) {
-			struct sockaddr *addr = &addr_ipv6;
-			socklen_t addrlen = sizeof(struct sockaddr);
 
 			if ((fds[i].revents & ZSOCK_POLLERR) ||
 			    (fds[i].revents & ZSOCK_POLLNVAL)) {
@@ -252,52 +273,58 @@ static void tcp_server_session(void)
 			}
 
 			switch (i) {
-			case SOCK_ID_IPV4_LISTEN:
-				addr = &addr_ipv4;
-				__fallthrough;
-			case SOCK_ID_IPV6_LISTEN:{
-				int sock = zsock_accept(fds[i].fd, addr, &addrlen);
+			case SOCK_ID_IPV4_LISTEN ... SOCK_ID_IPV6_LISTEN: {
+				int j = SOCK_ID_IPV6_LISTEN + 1;
+				struct sockaddr addr_incoming_conn;
+				socklen_t addrlen = sizeof(struct sockaddr);
+				int sock = zsock_accept(fds[i].fd,
+							&addr_incoming_conn,
+							&addrlen);
 
 				if (sock < 0) {
 					NET_ERR("TCP receiver IPv%d accept error",
-						(i <= SOCK_ID_IPV4_DATA) ? 4 : 6);
+						(sock_addr[i].sa_family == AF_INET
+							? 4 : 6));
 					goto error;
 				}
 
-				if (i == SOCK_ID_IPV4_LISTEN &&
-				    fds[SOCK_ID_IPV4_DATA].fd < 0) {
-					fds[SOCK_ID_IPV4_DATA].fd = sock;
-					fds[SOCK_ID_IPV4_DATA].events = ZSOCK_POLLIN;
-				} else if (i == SOCK_ID_IPV6_LISTEN &&
-					   fds[SOCK_ID_IPV6_DATA].fd < 0) {
-					fds[SOCK_ID_IPV6_DATA].fd = sock;
-					fds[SOCK_ID_IPV6_DATA].events = ZSOCK_POLLIN;
-				} else {
+				for (; j < SOCK_ID_MAX; j++) {
+					if (fds[j].fd < 0) {
+						break;
+					}
+				}
+
+				if (j == SOCK_ID_MAX) {
 					/* Too many connections. */
 					zsock_close(sock);
 					break;
 				}
 
+				fds[j].fd = sock;
+				fds[j].events = ZSOCK_POLLIN;
+				memcpy(&sock_addr[j], &addr_incoming_conn,
+						addrlen);
+
 				break;
 			}
 
-			case SOCK_ID_IPV4_DATA:
-				addr = &addr_ipv4;
-				__fallthrough;
-			case SOCK_ID_IPV6_DATA:
+			case (SOCK_ID_IPV6_LISTEN + 1) ... (SOCK_ID_MAX - 1):
 				ret = zsock_recv(fds[i].fd, buf, sizeof(buf), 0);
 				if (ret < 0) {
 					NET_ERR("recv failed on IPv%d socket (%d)",
-						(i <= SOCK_ID_IPV4_DATA) ? 4 : 6,
+						(sock_addr[i].sa_family == AF_INET
+							? 4 : 6),
 						errno);
-					goto error;
+					/* This will close the zperf session */
+					ret = 0;
 				}
 
-				tcp_received(addr, ret);
+				tcp_received(&sock_addr[i], ret);
 
 				if (ret == 0) {
 					zsock_close(fds[i].fd);
 					fds[i].fd = -1;
+					memset(&sock_addr[i], 0, sizeof(struct sockaddr));
 				}
 
 				break;
@@ -315,6 +342,7 @@ cleanup:
 	for (int i = 0; i < ARRAY_SIZE(fds); i++) {
 		if (fds[i].fd >= 0) {
 			zsock_close(fds[i].fd);
+			memset(&sock_addr[i], 0, sizeof(struct sockaddr));
 		}
 	}
 }
-- 
2.39.1

